server/server.go

```diff
@@ -127,113 +120,91 @@ func (s *Server) HandleEvent(ctx context.Context, e lookout.Event) error {
+		}
+	}
+	return err
+}
 
+func (s *Server) analyze(ctx context.Context, e lookout.Event, safePosting bool) ([]lookout.AnalyzerComments, error) {
+	ctxlog.Get(ctx).Infof("processing event type %d", e.Type())
+
+	var comments []lookout.AnalyzerComments
 	if err := e.Validate(); err != nil {
-		return err
+		return comments, err
 	}
 
 	conf, err := s.getConfig(ctx, e)
 	if err != nil {
-		return err
+		return comments, err
 	}
 
 	s.status(ctx, e, lookout.PendingAnalysisStatus)
 
-	send := func(
-		ctx context.Context,
-		a lookout.AnalyzerClient,
-		settings map[string]interface{},
-	) ([]*lookout.Comment, error) {
-		st := grpchelper.ToPBStruct(settings)
-		if st != nil {
-			e.Configuration = *st
-		}
-
-		if s.analyzerReviewTimeout > 0 {
-			var cancel context.CancelFunc
-			ctx, cancel = context.WithTimeout(ctx, s.analyzerReviewTimeout)
-			defer cancel()
-		}
-
-		resp, err := a.NotifyReviewEvent(ctx, e)
-		if err != nil {
-			return nil, err
-		}
-		return resp.Comments, nil
-	}
-	comments, err := s.concurrentRequest(ctx, conf, send, grpcErrorMessages[pb.ReviewEventType])
-	if err != nil {
-		return err
-	}
-
-	if err := s.post(ctx, e, comments, safePosting); err != nil {
-		s.status(ctx, e, lookout.ErrorAnalysisStatus)
-		return fmt.Errorf("posting analysis failed: %s", err)
-	}
-
-	s.status(ctx, e, lookout.SuccessAnalysisStatus)
-
-	return nil
+	send := s.getSender(e)
+	return s.concurrentRequest(ctx, conf, send, grpcErrorMessages[pb.ReviewEventType])
 }
 
-// HandlePush sends request to analyzers concurrently
-func (s *Server) HandlePush(ctx context.Context, e *lookout.PushEvent, safePosting bool) error {
-	ctx, logger := ctxlog.WithLogFields(ctx, log.Fields{
-		"provider": e.Provider,
-	})
-	logger.Infof("processing push")
-
-	if err := e.Validate(); err != nil {
-		return err
+func (s *Server) getSender(e lookout.Event) reqSent {
+	type analyzerNotifier struct {
+		notify  func(context.Context, lookout.AnalyzerClient) (*pb.EventResponse, error)
+		timeout time.Duration
 	}
 
-	conf, err := s.getConfig(ctx, e)
-	if err != nil {
-		return err
+	getAnalyzerNotifier := func(ctx context.Context, settings map[string]interface{}) (analyzerNotifier, error) {
+		st := grpchelper.ToPBStruct(settings)
+		switch ev := e.(type) {
+		case *lookout.ReviewEvent:
+			if st != nil {
+				ev.Configuration = *st
+			}
+			return analyzerNotifier{
+				func(ctx context.Context, a lookout.AnalyzerClient) (*pb.EventResponse, error) {
+					return a.NotifyReviewEvent(ctx, ev)
+				}, s.analyzerReviewTimeout}, nil
+		case *lookout.PushEvent:
+			if st != nil {
+				ev.Configuration = *st
+			}
+			return analyzerNotifier{
+				func(ctx context.Context, a lookout.AnalyzerClient) (*pb.EventResponse, error) {
+					return a.NotifyPushEvent(ctx, ev)
+				}, s.analyzerPushTimeout}, nil
+		default:
+			ctxlog.Get(ctx).Debugf("ignoring unsupported event: %s", ev)
+			return analyzerNotifier{}, fmt.Errorf("unsupported event: %s", ev)
```

Comment:
I might be mistaken, but this PR is also changing the `s.HandleEvent(ctx, strangeEvent)` behavior when it is sent an event that was not a `ReviewEvent` nor a `PushEvent`;
If I'm rightm the `s.HandleEvent(ctx, strangeEvent)` will fail with an `unsupported` error message, instead of just logging a debug `unsupported` message.

If that's true, I think we should not change the behavior in a refactor PR or without a clear reasons to do so.

Reply:
You're right. With these changes the `HandleEvent` method fails in case on supported event. Another thing to note with this refactor is that in case of unsupported event the check is so deep that it even spawns a goroutine for each analyzer. And this is a big no-no in my opionion.